packages/enhanced/src/lib/sharing/ConsumeSharedPlugin.ts

```diff
@@ -296,33 +294,69 @@ class ConsumeSharedPlugin {
 
         normalModuleFactory.hooks.factorize.tapPromise(
           PLUGIN_NAME,
-          ({ context, request, dependencies }) =>
+          async (resolveData: ResolveData): Promise<Module | undefined> => {
+            const { context, request, dependencies, contextInfo } = resolveData;
             // wait for resolving to be complete
             //@ts-ignore
-            promise.then(() => {
+            return promise.then(() => {
               if (
                 dependencies[0] instanceof ConsumeSharedFallbackDependency ||
                 dependencies[0] instanceof ProvideForSharedDependency
               ) {
                 return;
               }
+
+              // First try to match with layer-specific request
+              if (contextInfo.issuerLayer) {
+                // Try to find a layer-specific match
+                for (const [key, options] of unresolvedConsumes) {
```

Comment:
This should be handled in the `unresolvedConsumes.get` case below

Reply:
Ive opened up a new PR with additional adjustments for this, it becomes hard to control due to how request is used.
I propose adding another option explicitly for controlling the request matching.